src/Worker/Core/Shims/TaskOrchestrationShim.cs

```diff
@@ -46,10 +46,19 @@ partial class TaskOrchestrationShim : TaskOrchestration
 
         object? input = this.DataConverter.Deserialize(rawInput, this.implementation.InputType);
         this.wrapperContext = new(innerContext, this.invocationContext, input);
-        object? output = await this.implementation.RunAsync(this.wrapperContext, input);
 
-        // Return the output (if any) as a serialized string.
-        return this.DataConverter.Serialize(output);
+        try
+        {
+            object? output = await this.implementation.RunAsync(this.wrapperContext, input);
+
+            // Return the output (if any) as a serialized string.
+            return this.DataConverter.Serialize(output);
+        }
+        finally
+        {
+            // if user code crashed inside a critical section, or did not exit it, do that now
+            this.wrapperContext.ExitCriticalSectionIfNeeded();
```

Comment:
This works _for now_. But not ideal that this logic only happens as part of the wrapper shim. Fine for now I suppose.

Reply:
Yeah, note that there are already some paths on which the lock fails to release (e.g. under nondeterministic orchestration errors). It is not completely avoidable unfortunately.

That's why we have the 'release orphaned locks' as a backup "solution".